src/calendarimportmodel.cpp

```diff
@@ -187,66 +173,38 @@ QHash<int, QByteArray> CalendarImportModel::roleNames() const
     roleNames[EndTimeRole] = "endTime";
     roleNames[AllDayRole] = "allDay";
     roleNames[LocationRole] = "location";
+    roleNames[DuplicateRole] = "duplicate";
     roleNames[UidRole] = "uid";
 
     return roleNames;
 }
 
-static bool incidenceLessThan(const KCalendarCore::Incidence::Ptr e1,
-                              const KCalendarCore::Incidence::Ptr e2)
-{
-    if (e1->dtStart() == e2->dtStart()) {
-        int cmp = QString::compare(e1->summary(),
-                                   e2->summary(),
-                                   Qt::CaseInsensitive);
-        if (cmp == 0)
-            return QString::compare(e1->uid(), e2->uid()) < 0;
-        else
-            return cmp < 0;
-    } else {
-        return e1->dtStart() < e2->dtStart();
-    }
-}
-
-void CalendarImportModel::reload()
-{
-    if (!mFileName.isEmpty() || !mIcsRawData.isEmpty()) {
-        bool success = importToMemory(mFileName, mIcsRawData);
-        setError(!success);
-    } else if (!mEventList.isEmpty()) {
-        beginResetModel();
-        mEventList.clear();
-        endResetModel();
-        emit countChanged();
-        setError(false);
-    }
-}
-
 bool CalendarImportModel::importToMemory(const QString &fileName, const QByteArray &icsData)
 {
-    if (!mEventList.isEmpty())
-        mEventList.clear();
-
     bool success = false;
 
-    beginResetModel();
     KCalendarCore::MemoryCalendar::Ptr cal(new KCalendarCore::MemoryCalendar(QTimeZone::systemTimeZone()));
     if (!fileName.isEmpty()) {
         success = CalendarUtils::importFromFile(fileName, cal);
-    } else {
+    } else if (!icsData.isEmpty()) {
         success = CalendarUtils::importFromIcsRawData(icsData, cal);
     }
-    KCalendarCore::Incidence::List incidenceList = cal->incidences();
-    for (int i = 0; i < incidenceList.size(); i++) {
-        KCalendarCore::Incidence::Ptr incidence = incidenceList.at(i);
-        if (incidence->type() == KCalendarCore::IncidenceBase::TypeEvent)
-            mEventList.append(incidence.staticCast<KCalendarCore::Event>());
-    }
-    if (!mEventList.isEmpty())
-        qSort(mEventList.begin(), mEventList.end(), incidenceLessThan);
 
+    beginResetModel();
+    mEventList = cal->events(KCalendarCore::EventSortStartDate);
+    // To avoid detach here, use qAsConst when available.
+    for (const KCalendarCore::Event::Ptr &event : mEventList) {
+        mStorage->load(event->uid(), event->recurrenceId());
+        const KCalendarCore::Event::Ptr old =
+            mStorage->calendar()->event(event->uid(), event->recurrenceId());
+        if (old) {
+            mDuplicates.insert(old->instanceIdentifier());
```

Comment:
mDuplicated should be emptied somewhere in this method?

Reply:
Good catch, I'll update the MR. Thanks.